File: stacks&queues/71. Simplify Path.js
// TC: O(N) SC: O(N)

/**
 * @param {string} path
 * @return {string}
 */
var simplifyPath = function(path) {
  let link = path.split("/").filter(el => el != "" && el != ".")
  // console.log(link)
  let stack = []
  for (let i of link){
      if (i == ".."){
        stack.pop()  
      }
      else stack.push(i)
      
  }
  
  return "/" + stack.join("/")
};

File: stacks&queues/71. Simplify Path.py
#TC: O(N) SC: O(N)

# key takeaway: popping an empty array throws error python while it will simply return undefined in JS(sneaky and pernicious)

class Solution:
    def simplifyPath(self, path: str) -> str:
        link = path.split("/")
        stack = []
        for ele in link:
            if ele == "" or ele == ".":
                continue
            elif ele == ".." :
                if stack:
                    stack.pop()
            else:
                stack.append(ele)
                
        return "/" + "/".join(stack)